Add Sphere Mask and Particles Turbulence Nodes and Posterize ColorFunc

paddy-exe opened the pull request:
Adds Posterize ColorFunc Mode to CanvasItem and Spatial Visual Shader Nodes 
Adds Particle Turbulence to Particle Visual Shader Nodes

Closes https://github.com/godotengine/godot-proposals/issues/5106
Part of https://github.com/godotengine/godot-proposals/issues/7666

## Showcase
### Posterize node
![grafik](https://github.com/godotengine/godot/assets/38077837/4e80cb5d-45ad-4ce8-a834-353135fad26e)

### ParticleTurbulence

https://github.com/godotengine/godot/assets/38077837/23ea529b-ff6b-4d2c-90b9-c6edbd64f858

_Production edit: closes godotengine/godot-roadmap#7_

AThousandShips commented on scene/resources/visual_shader_particle_nodes.cpp:
```diff
@@ -1162,6 +1162,205 @@ VisualShaderNodeParticleAccelerator::VisualShaderNodeParticleAccelerator() {
+String VisualShaderNodeParticleTurbulence::generate_global_per_node(Shader::Mode p_mode, int p_id) const {
+	String code;
+
+	code += "vec4 turb_grad(vec4 p) {\n";
```
I'd suggest doing this with R strings

paddy-exe replied:
Could you give me an example or point me to a file where it is implemented with R strings? Haven't used R strings yet

AThousandShips replied:
See `sky_material` for an example 🙂

paddy-exe replied:
Wait, you mean for `generate_global_per_node`? Why there? If anywhere I would assume `generate_code`

AThousandShips replied:
Because it's a long block of strings being added and it's much cleaner if it's in one block of string instead

paddy-exe replied:
Should be done now :D

Geometror replied:
Allowed myself to mark this as resolved :)

AThousandShips commented on scene/resources/visual_shader_nodes.cpp:
```diff
@@ -3179,15 +3179,46 @@ String VisualShaderNodeColorFunc::get_caption() const {
+				case 1:
+					return "Steps";
+				default:
+					break;
+			}
+			break;
+		default:
+			return "";
+	}
```
```suggestion
					break;
			}
			break;
		default:
			break;
	}
	return "";
```

AThousandShips commented on scene/resources/visual_shader_nodes.cpp:
```diff
@@ -3179,15 +3179,46 @@ String VisualShaderNodeColorFunc::get_caption() const {
+			}
+			break;
+		default:
+			return PORT_TYPE_VECTOR_3D;
+	}
```
```suggestion
			break;
	}
	return PORT_TYPE_VECTOR_3D;
```

Geometror commented on doc/classes/VisualShaderNodeParticleTurbulence.xml:
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<class name="VisualShaderNodeParticleTurbulence" inherits="VisualShaderNode" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../class.xsd">
+	<brief_description>
+		Visual shader node for randomized particle movement.
```
For consistency :)
```suggestion
		A visual shader node for randomized particle movement.
```

Geometror commented on scene/resources/visual_shader_particle_nodes.cpp:
```diff
@@ -1162,6 +1162,206 @@ VisualShaderNodeParticleAccelerator::VisualShaderNodeParticleAccelerator() {
+	float s1 = turb_noise(p + vec4(vec3(0.0), 1.7320508 * 2048.333333));
+	float s2 = turb_noise(p - vec4(vec3(0.0), 1.7320508 * 2048.333333));
+	vec3 c = vec3(s, s1, s2);
+	return c;
+}
+
+vec3 turb_curl_3d(vec4 p, float c) {
+	float epsilon = 0.001 + c;
+	vec4 dx = vec4(epsilon, 0.0, 0.0, 0.0);
+	vec4 dy = vec4(0.0, epsilon, 0.0, 0.0);
+	vec4 dz = vec4(0.0, 0.0, epsilon, 0.0);
+	vec3 x0 = turb_noise_3x(p - dx).xyz;
+	vec3 x1 = turb_noise_3x(p + dx).xyz;
+	vec3 y0 = turb_noise_3x(p - dy).xyz;
+	vec3 y1 = turb_noise_3x(p + dy).xyz;
+	vec3 z0 = turb_noise_3x(p - dz).xyz;
+	vec3 z1 = turb_noise_3x(p + dz).xyz;
+	float x = (y1.z - y0.z) - (z1.y - z0.y);
+	float y = (z1.x - z0.x) - (x1.z - x0.z);
+	float z = (x1.y - x0.y) - (y1.x - y0.x);
+	return normalize(vec3(x, y, z));
+}
+
+vec3 get_turb_noise_direction(vec3 __pos, float __turb_n_strength, float __turb_n_scale, vec3 __turb_n_speed, float __turb_n_speed_random) {
+	float __adj_contrast = max((__turb_n_strength - 1.0), 0.0) * 70.0;
+	vec4 __noise_time = TIME * vec4(__turb_n_speed, __turb_n_speed_random);
+	vec4 __noise_pos = vec4(__pos * __turb_n_scale, 0.0);
+	vec3 __noise_direction = turb_curl_3d(__noise_pos + __noise_time, __adj_contrast);
+	__noise_direction = mix(0.9 * __noise_direction, __noise_direction, __turb_n_strength - 9.0);
+	return __noise_direction;
+}
+
+float get_turb_influence(uint __noise_seed, float __turb_influence_min, float __turb_influence_max, float __turb_influence_over_life) {
+	float __turb_influence_range = mix(__turb_influence_min, __turb_influence_max, __rand_from_seed(__noise_seed));
+	float __turb_influence = clamp(__turb_influence_range * __turb_influence_over_life, 0.0, 1.0);
+	return __turb_influence;
+}
+	)");
+
+	return code;
```
Do you think it would be feasible/make sense to share this with `ParticleProcessMaterial` to avoid duplication?

paddy-exe replied:
I don't think it makes sense here to combine these because I had to change quite a few parts of the shader code to make it usable for visual shaders due to the separation of the `start` and `process` functions in the particle shader. Furthermore, the `ParticleProcessMaterial` can't be converted to a visual shader anyway so I currently see no reason to do so.

Geometror commented on doc/classes/VisualShaderNodeParticleTurbulence.xml:
```diff
@@ -0,0 +1,11 @@
+		Visual shader node for randomized particle movement.
+	</brief_description>
+	<description>
+		The Turbulence Node will output a noise direction and turbulence influence values based on the given noise parameters which can be used to influence the give particles random velocity.
```
```suggestion
		The Turbulence Node generates a noise direction and turbulence influence value based on the given noise parameters.
		These can be used to add a turbulent motion to the particle's velocity, which allows for creating organic and natural-looking particle effects.
```

Geometror commented on doc/classes/VisualShaderNodeSphereMask.xml:
```diff
@@ -0,0 +1,11 @@
+		A visual shader node that smoothly masks via multiplication.
+	</brief_description>
+	<description>
+		The SphereMask Node offers circular masking options with optional harshness and center specifications. It can be used to blend out the outer edges of a sprite.
```
Did you mean "hardness"? :)

Geometror reviewed (commented):
Posterize needs its own AddOption (`visual_shader_editor_plugin.cpp` ~line 6210) so that you can add it directly like the other color functions.

I played a bit with every node and it looks like everything is working fine.

Ideally this PR would be split up since (I think) these nodes are unrelated to each other, but nonetheless great work!

paddy-exe commented:
@Geometror Thanks for the detailed feedback! I have amended the changes you suggested

Ansraer reviewed (changes requested):
The sphere mask node sounds like a useful node to have (I know that I created something similar it manually multiple times in the past) but I really think the name should be changed.
"Mask" kind of implies for me that this can only be used to mask an existing texture, which is only one usecase of it. In addition the node itself doesn't mask anything, needing an additional Multiplication OP node for that purpose.

IMO "SphereShape" would be a far better name that better communicates what this node actually does.

Geometror commented on scene/resources/visual_shader_particle_nodes.cpp:
```diff
@@ -1162,6 +1162,206 @@ VisualShaderNodeParticleAccelerator::VisualShaderNodeParticleAccelerator() {
+}
+
+VisualShaderNodeParticleTurbulence::VisualShaderNodeParticleTurbulence() {
+	set_input_port_default_value(2, 1.0); // noise strength
+	set_input_port_default_value(3, 9.0); // noise scale
+	set_input_port_default_value(4, Vector3(0.0, 0.0, 0.0)); // noise speed
+	set_input_port_default_value(5, 0.2); // noise speed random
+	set_input_port_default_value(6, 0.1); // influence min
+	set_input_port_default_value(7, 0.1); // influence max
+	set_input_port_default_value(8, 1.0); // influence over life
```
I know this is the way we currently do it for almost all nodes, but we should really start to use enums to describe the ports/slots. (not a blocker, just for reference)

Geometror reviewed (commented):
@Ansraer I have to disagree here, in my opinion "mask" is a better term to describe what it does (and is likely the most common usecase, not just to mask textures, but whole parts of shaders). However, I'm fine with "shape" too.
But I think it should be called CircleMask since it's 2D.

Geometror commented:
Besides that, I think it would be of value to implement it as a general Shape/Mask node (which allows selecting a shape using a dropdown menu). Example from material maker:
![grafik](https://github.com/godotengine/godot/assets/50084500/a6c93c8c-1fc1-4f9b-a8c6-64e5d7673427)
Well, I just noticed it's called "Shape" there, one point for Ansraer :)

paddy-exe committed:
Add Particles Turbulence Nodes and Posterize ColorFunc

Adds Posterize ColorFunc Mode to CanvasItem and Spatial Visual Shader Nodes
Adds Particle Turbulence to Particle Visual Shader Nodes

paddy-exe commented:
Updated the PR and removed the SphereMask node for now. I have the changes locally still.
